### PR TITLE
[MRG] FIX: Support commas in BrainVision channel names, event types, and event descriptions

### PR DESCRIPTION
**Pull request by sappelhoff**
This came up in https://github.com/bids-standard/pybv/pull/53

In the BrainVision format channel names, event types, and event descriptions that contain a comma `,` are encoded such that the comma is written as `\1`.

I add a test here to show that MNE-Python does currently not adequately replace the `\1` characters to `,` during reading.

This is potentially broken for event type and event description as well, but I haven't looked into that yet.

Should we fix it?

### CODE REVIEW

File: mne/io/brainvision/tests/test_brainvision.py
Comment (larsoner) [inline mne/io/brainvision/tests/test_brainvision.py:297]: no need for this, `pytest` should clean up on its own

**Review by larsoner — COMMENTED** [review]
> Should we fix it?

Yes I think fixing the channel names and the event type and descriptions would be good

File: mne/io/brainvision/tests/test_brainvision.py
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:297]: even without a fixture or anything?

Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py:297]: Oh sorry I assumed you were using a `tmpdir` fixture. You definitely should use `tmpdir` and `shutil.copyfile` instead of writing anything to the `vhdr_path` directory

File: mne/io/brainvision/tests/test_brainvision.py
Comment (sappelhoff) [inline mne/io/brainvision/tests/test_brainvision.py:687]: locally, this test failed, and I have no idea why.

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:687]: that is --> it does not fail because of my cosmetic change here, but because original_event_id and event_id are not equal. See:

```
(Pdb) event_id
{'Comma,Type/CommaValue,1': 1, 'Event/254': 2, 'New Segment/': 3, 'Optic/O  1': 4, 'Response/R255': 5, 'Stimulus/S253': 6, 'Stimulus/S255': 7, 'SyncStatus/Sync On': 8}
(Pdb) original_event_id
{'Comma,Type/CommaValue,1': 10001, 'Event/254': 254, 'New Segment/': 99999, 'Optic/O  1': 2001, 'Response/R255': 1255, 'Stimulus/S253': 253, 'Stimulus/S255': 255, 'SyncStatus/Sync On': 99998}
(Pdb) 
```

Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py:687]: It somehow has to do with your changes to `test.vmrk` above. Did you mean to revert that?

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:687]: No, that is a permanent change I wanted to add to the test file :thinking: but looking at the code, I also don't see why that would be a problem:

```Python
    fname = tmpdir / 'bv-raw.fif'
    raw = read_raw_brainvision(vhdr_path, eog=eog)
    original_events, original_event_id = events_from_annotations(raw)

    raw.save(fname)
    raw = read_raw_fif(fname)
    events, event_id = events_from_annotations(raw)

    assert event_id == original_event_id
    assert_array_equal(events, original_events)
```

1. we read the BV file
1. we get events from it (which works - even for the newly added event, as shown in the other test)
1. we save it to FIF
1. we read back from FIF
1. we get events from the just-read FIF annotations
1. ... they are different?

Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py:687]: Probably based on `event_id='auto'` default and `_check_bv_annot`, somehow your new event makes `_check_bv_annot` return False rather than True

Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py:687]: ... to avoid needing to deal with this in this PR, I would do the same trick as above, namely don't modify the actual `test.mrk` file, but copy it line-by-line and add an event at the bottom

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:687]: ahhh got it @larsoner ! Because of these lines:

https://github.com/mne-tools/mne-python/blob/fa7064a8d4cf8857d66ca001a17fa66e7121d3bc/mne/annotations.py#L956-L963

basically, the first raw is a "brainvision raw" ... and the second raw is a "fif raw" ... and the annotation-reading with event_id="auto" depends on the kind of raw format.

So passing an event_id other than "auto" at the second events_from_annotations call should fix it.

But is that what we want? The test confuses me a bit with it's goal.

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:687]: yup, `events, event_id = events_from_annotations(raw, original_event_id)` fixes it

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:687]: oh weird, I didn't see your comments :) should have refreshed to save me some detective work

Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py:687]: Don't do that. The goal of the test is that "standard BV events get parsed as BV events regardless of whether it's RawBrainvision or RawFIF". Passing the event_id defeats this purpose. Hence my suggestion to use the `tmpdir` + rewrite + append approach in the one test where you actually want to check for comma support rather than modifying `test.mrk` on disk directly in this PR

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:687]: I see, and introducing non-standard events obviously breaks the thing, reasonably so. Thanks!

**Comment by sappelhoff** [comment]
Would this PR be eligible for a backport? It may come in handy for the pybv package.

**Comment by larsoner** [comment]
Sure but we just cut 0.21.1 yesterday so it might be a few weeks or so until we do a 0.21.2

**Review by agramfort — APPROVED** [review]

**Review by larsoner — APPROVED** [review]

File: mne/io/brainvision/tests/test_brainvision.py
Comment (sappelhoff) [inline mne/io/brainvision/tests/test_brainvision.py:301]: The tests seem to be failing on Windows. Perhaps the manipulation of the test file prior to the test run is not working correctly on Windows?

Maybe this suggestion can help us to be sure:

```suggestion
    new_lines = []
    nperformed_replacements = 0
    for line in lines:
        for to_replace, replacement in replace_dict.items():
            if to_replace in line:
                new = line.replace(to_replace, replacement)
                new_lines.append(new)
                nperformed_replacements += 1
                break
        else:
            new_lines.append(line)
    assert nperformed_replacements == len(replace_dict)
```

**Comment by sappelhoff** [comment]
indeed, on Windows, the temporary file-replacement for testing is not working as expected:

The windows tests only fail with:

```
mne\io\brainvision\tests\test_brainvision.py:304: in test_ch_names_comma
    assert nperformed_replacements == len(replace_dict)
E   assert 1 == 2
E     +1
E     -2
```

**Comment by sappelhoff** [comment]
@cbrnr could you please have a look at this one and merge if you are happy? 

Then I can switch on the corresponding test on pybv with `@requires_version("mne", "0.22")`

File: doc/changes/latest.inc
Comment (cbrnr) [inline doc/changes/latest.inc:108]:
```suggestion
- Fix bug where BrainVision channel names, event types, and event descriptions containing commas were incorrectly parsed (:gh:`8492` by `Stefan Appelhoff`_)
```

Comment (cbrnr) [reply doc/changes/latest.inc:108]: See #8442.

Comment (sappelhoff) [reply doc/changes/latest.inc]: done!

File: mne/io/brainvision/tests/test_brainvision.py
Comment (cbrnr) [inline mne/io/brainvision/tests/test_brainvision.py:631]: Why the extra line?

**Review by cbrnr — CHANGES_REQUESTED** [review]

File: mne/io/brainvision/tests/test_brainvision.py
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:309]: indeed, thanks for the catch! fixed in f3ac8a1e7bdbb53ad4eade8ee6a999727bbad1d8

File: mne/io/brainvision/tests/test_brainvision.py
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:631]: to end a file with an empty new line, which is the default on Linux and MacOS, but strictly speaking it would not be necessary in this case.

Comment (cbrnr) [reply mne/io/brainvision/tests/test_brainvision.py:631]: No, I think the `\n` could be combined with the previous line.

File: mne/io/brainvision/tests/test_brainvision.py
Comment (cbrnr) [inline mne/io/brainvision/tests/test_brainvision.py:281]: `r"4     F4,foo "` doesn't need to be a raw string.

Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py:281]:
```suggestion
        r"^4\s\s\s\s\sF4": "4     F4,foo ",
```

File: mne/io/brainvision/tests/test_brainvision.py
Comment (sappelhoff) [inline mne/io/brainvision/tests/test_brainvision.py:631]:
```suggestion
        fout.write(r"Mk15=Comma\1Type,CommaValue\11,7800,1,0\n")
```

**Review by cbrnr — APPROVED** [review]

**Comment by cbrnr** [comment]
I'll merge once CIs come back green.

**Comment by cbrnr** [comment]
Not sure what's up with the [failing test](https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=10040&view=logs&j=c0130b29-789d-5a3c-6978-10796a508a7f&t=b81cfa0d-2d5e-5a59-e316-838f621b71a0). It stopped after `mne/tests/test_report.py::test_render_add_sections[mayavi] PASSED        [ 75%]`, but it doesn't say which test failed and why.

**Comment by sappelhoff** [comment]
I don't know either. 🤷‍♂️ but it passed before and it passes locally

**Comment by cbrnr** [comment]
I'll re-run and see if this magically resolves the issue (it often does).

**Comment by cbrnr** [comment]
Now Travis fails at the same test (`mne/tests/test_report.py::test_render_add_sections[pyvista] Traceback (most recent call last):`)? I'd like to clarify this before merging. @larsoner do you have any idea?

**Comment by GuillaumeFavelier** [comment]
> Now Travis fails at the same test

It's related to https://github.com/mne-tools/mne-python/pull/8494

I'm working on a fix.

**Comment by sappelhoff** [comment]
@cbrnr I rebased on Guillaume's fix and CIs pass now :-)

**Comment by cbrnr** [comment]
Thanks @sappelhoff!

**Pull request merged**
